nixos/darkhttpd: change listening address default to ::1

vojta001 opened the pull request:
We unconditionally pass --ipv6 to the binary. In older versions, setting both --addr 127.0.0.1 and --ipv6 resulted in listening on [::], a potential security issue. On the other hand, the binary started.

With the current version however, --ipv6 implies an IPv6 address in --addr resulting in a crash with: "darkhttpd: malformed --addr argument"

While this commit is technically a BC-break, it should be harmless as existing users either do not explicitly set the address and the server refuses the start (making this change rather a fix) or they set the address and are therefore unaffected.

----

The above excerpt from the commit message shall summarize the changes well. I just want to add that the _this is not a BC-break, this is a bug fix_ shall qualify it for merging despite the freeze in #339153.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

vojta001 committed:
nixos/darkhttpd: change listening address default to ::1

We unconditionally pass --ipv6 to the binary. In older versions, setting
both --addr 127.0.0.1 and --ipv6 resulted in listening on [::],
a potential security issue. On the other hand, the binary started.

With the current version however, --ipv6 implies an IPv6 address in
--addr resulting in a crash with: "darkhttpd: malformed --addr argument"

While this commit is technically a BC-break, it should be harmless as
existing users either do not explicitly set the address and the server
refuses the start (making this change rather a fix) or they set the
address and are therefore unaffected.

patka-123 reviewed (approved):
Makes sense and LGTM!

vojta001 committed:
nixos/darkhttpd: fix comment on how to listen on all interfaces

This does not seem to have ever been intended, it was probably caused by
the same parsing issue as described in commit abdadde8a2fc4a5a2643a81eba97e39301efcaae

Listening on [::] makes darkhttpd see IPv4 connections under ::ffff:a.b.c.d

bobvanderlinden reviewed (approved):
Makes sense as a fix for the most common behavior (ipv6 enabled).

Machines without ipv6 do currently work correct? The PR change will break such setups with a default binding?

It seems the default should ideally be dependent on the enableipv6 option?

SuperSandro2000 commented on nixos/modules/services/web-servers/darkhttpd.nix:
```diff
@@ -27,11 +27,11 @@ in {
 
     address = mkOption {
-      default = "127.0.0.1";
+      default = "::1";
```
```suggestion
      default = if config.networking.enableIPv6 then "::1" else "127.0.0.1";
      defaultText = lib.literalExpresison ''if config.networking.enableIPv6 then "::1" else "127.0.0.1"'';
```

FliegendeWurst replied:
Then we would also have to adjust:

> We unconditionally pass --ipv6 to the binary

SuperSandro2000 replied:
🤔 probably, I am unsure

vojta001 commented:
@bobvanderlinden @SuperSandro2000 Oh I completely forgot about machines without IPv6. In such a case I don't think we should detect IPv6 status by the `config.networking.enableIPv6` option. It's good for setting defaults, but not for deciding the flag `--ipv6` as it makes the binary crash in case you disable IPv6 via other means (I can think of toggling some systemd options for this unit only).

I therefore propose to add another config options for darkhttpd to decide whether to enable IPv6 support. It would default to `config.networking.enableIPv6` and control the other stuff discussed here
